source/client-side-encryption/client-side-encryption.rst

```diff
@@ -286,13 +289,45 @@ Data keys are stored as documents in a special MongoDB collection. Data
 keys are protected with encryption by a KMS provider (AWS KMS, Azure key
 vault, GCP KMS, or a local master key).
 
+metadataClient
+^^^^^^^^^^^^^^
+A client used to run ``listCollections`` to determine whether a collection has
+an associated JSON schema with encrypted fields specified.
+
+If a ``metadataClient`` is not set, and ``bypassAutomaticEncryption=false`` then
```

Comment:
From reading this section several times, I'm still not sure I understand exactly what the configuration will be in every case:

1. metadataClient is passed in by user, bypassAutomaticEncryption=false
2. metadataClient is passed in by user, bypassAutomaticEncryption=true
3. metadataClient is not passed in by user, bypassAutomaticEncryption=false
4. metadataClient is not passed in by user, bypassAutomaticEncryption=true

As best as I can tell, the first two cases are the same, i.e. the metadataClient is used for running `listCollections`. In case 3, it sounds like the driver is supposed to create a new client interally and set `metadataClient` to that. I'm not sure about the fourth case; does `bypassAutomaticEncryption` mean that `listCollections` won't occur?

Either way, only one of the four cases is explicitly defined here (case 3), and the other cases are left implicit. One thing that's not clear is if the driver is supposed to actually populate `metadataClient` in the case that none is passed in or if it should just substitute `internal_client` at the places where `metadataClient` would normally be used. If the former, I think the language here should be updated to "If a `metadataClient` is not passed in..." (rather than "is not set"), and if the latter, think the docstring on `metadataClient` above should be revised to be clear that the value is not actually set to the value of internal_client.

Reply:
Great points, I think enumerating the cases is helpful. I described this behavior with pseudo-code in a separate section to show the configuration behavior with `metadataClient`, `keyVaultClient`, and the internal client. Hopefully that clarifies.

> does bypassAutomaticEncryption mean that listCollections won't occur?

Yes, `listCollections` is only run for automatic encryption.

> One thing that's not clear is if the driver is supposed to actually populate metadataClient in the case that none is passed in or if it should just substitute internal_client at the places where metadataClient would normally be used. If the former, I think the language here should be updated to "If a metadataClient is not passed in..." (rather than "is not set"), and if the latter, think the docstring on metadataClient above should be revised to be clear that the value is not actually set to the value of internal_client.

The former, I've updated the language.